tools/yarn-cling/lib/hoisting.ts

```diff
@@ -2,50 +2,148 @@ import { PackageLockPackage } from './types';
-        moveChildrenUp(depPackage, parentContainer);
-      } else if (parentContainer[depName].version === depPackage.version) {
-        // Already exists, no conflict, delete the child, no need to recurse
-        delete parent.dependencies[depName];
-        didChange = true;
-      } else {
-        // There is a conflict, leave the second package where it is, but do recurse.
-        moveChildrenUp(depPackage, parent.dependencies);
+  function removeDupes(node: PackageLockPackage, rootPath: Array<PackageLockPackage>) {
+    if (!node.dependencies) { return; }
+
+    // Any dependencies here that are the same in the parent can be removed
+    for (const [depName, depPackage] of Object.entries(node.dependencies)) {
+      if (findInheritedDepVersion(depName, rootPath) === depPackage.version) {
+        delete node.dependencies[depName];
       }
     }
 
-    // Cleanup for nice printing
-    if (Object.keys(parent.dependencies).length === 0) {
-      delete parent.dependencies;
-      didChange = true;
+    // Recurse
+    for (const child of Object.values(node.dependencies)) {
+      removeDupes(child, [node, ...rootPath]);
     }
   }
+
+  function removeUseless(node: PackageLockPackage) {
+    if (!node.dependencies) { return; }
+    for (const [depName, depPkg] of Object.entries(node.dependencies)) {
+      if (!necessaryInTree(depName, depPkg.version, node)) {
+        delete node.dependencies[depName];
+      }
+    }
+
+    // Recurse
+    for (const child of Object.values(node.dependencies)) {
+      removeUseless(child);
+    }
+
+    // If we ended up with empty dependencies, just get rid of the key (for clean printing)
+    if (Object.keys(node.dependencies).length === 0) {
+      delete node.dependencies;
+    }
+  }
+
+  function findInheritedDepVersion(name: string, parentDependenciesChain: Array<PackageLockPackage>) {
+    for (const deps of parentDependenciesChain) {
+      if (deps.dependencies?.[name]) {
+        return deps.dependencies[name].version;
+      }
+    }
+    return undefined;
+  }
+
+  function recordOriginalDependencies(node: PackageLockPackage) {
+    if (!node.dependencies) { return; }
+
+    let list = originalDependencies.get(node);
+    if (!list) {
+      list = [];
+      originalDependencies.set(node, list);
+    }
+
+    for (const [depName, depPkg] of Object.entries(node.dependencies)) {
+      list.push(`${depName}@${depPkg.version}`);
+      recordOriginalDependencies(depPkg);
+    }
+  }
+
+  function necessaryInTree(name: string, version: string, tree: PackageLockPackage) {
+    if (originalDependencies.get(tree)?.includes(`${name}@${version}`)) {
+      return true;
+    }
+    if (!tree.dependencies) { return false; }
+
+    for (const depPackage of Object.values(tree.dependencies)) {
+      if (necessaryInTree(name, version, depPackage)) { return true; }
+    }
+    return false;
+  }
+}
+
+export function renderTree(tree: PackageLockPackage): string[] {
```

Comment:
If this is purely being used by the testing code, perhaps it should be moved to `hoisting.test.ts` (or a comment explaining why it's here).